client/js/src/client.ts

```diff
@@ -185,173 +229,339 @@ export async function client(
-						if (!evt.wasClean) {
-							fire_event({
-								type: "status",
-								status: "error",
-								message: BROKEN_CONNECTION_MSG,
-								queue: true,
-								endpoint,
-								fn_index
-							});
-						}
-					};
-
-					websocket.onmessage = function (event) {
-						const _data = JSON.parse(event.data);
-						const { type, status, data } = handle_message(
-							_data,
-							last_status[fn_index]
-						);
-
-						if (type === "update" && status) {
-							// call 'status' listeners
-							fire_event({ type: "status", endpoint, fn_index, ...status });
-							if (status.status === "error") {
 								websocket.close();
-								rej(status);
+							} else if (type === "generating") {
+								fire_event({
+									type: "status",
+									...status,
+									status: status?.status!,
+									queue: true,
+									endpoint,
+									fn_index
+								});
 							}
-						} else if (type === "hash") {
-							websocket.send(JSON.stringify({ fn_index, session_hash }));
-							return;
-						} else if (type === "data") {
-							websocket.send(JSON.stringify({ ...payload, session_hash }));
-						} else if (type === "complete") {
-							fire_event({
-								type: "status",
-								...status,
-								status: status?.status!,
-								queue: true,
-								endpoint,
-								fn_index
-							});
-							websocket.close();
-						} else if (type === "generating") {
-							fire_event({
-								type: "status",
-								...status,
-								status: status?.status!,
-								queue: true,
-								endpoint,
-								fn_index
-							});
-						}
-						if (data) {
-							fire_event({
-								type: "data",
-								data: data.data,
-								endpoint,
-								fn_index
-							});
-							res({ data: data.data });
-						}
-					};
+							if (data) {
+								fire_event({
+									type: "data",
+									data: data.data,
+									endpoint,
+									fn_index
+								});
+							}
+						};
+					}
 				}
-			});
+			);
+
+			function fire_event<K extends EventType>(event: Event<K>) {
+				const narrowed_listener_map: ListenerMap<K> = listener_map;
+				let listeners = narrowed_listener_map[event.type] || [];
+				listeners?.forEach((l) => l(event));
+			}
+
+			function on<K extends EventType>(
+				eventType: K,
+				listener: EventListener<K>
+			) {
+				const narrowed_listener_map: ListenerMap<K> = listener_map;
+				let listeners = narrowed_listener_map[eventType] || [];
+				narrowed_listener_map[eventType] = listeners;
+				listeners?.push(listener);
+
+				return { on, off, cancel };
+			}
+
+			function off<K extends EventType>(
+				eventType: K,
+				listener: EventListener<K>
+			) {
+				const narrowed_listener_map: ListenerMap<K> = listener_map;
+				let listeners = narrowed_listener_map[eventType] || [];
+				listeners = listeners?.filter((l) => l !== listener);
+				narrowed_listener_map[eventType] = listeners;
+
+				return { on, off, cancel };
+			}
+
+			function cancel() {
+				fire_event({
```

Comment:
In older versions of gradio, the iterator state will not be reset automatically when the websocket closes. So if you keep the session_hash the same, the next prediction would start off where the previous one left of which isn't great.

We manually reset the [iterator state](https://github.com/gradio-app/gradio/blob/main/client/python/gradio_client/utils.py#L225) by hitting the `/reset` route.